File: backend/governance/src/main/java/com/commonsengine/governance/service/GovernanceService.java
package com.commonsengine.governance.service;

import com.commonsengine.governance.domain.Model.Proposal;
import com.commonsengine.governance.domain.Model.ProposalId;
import com.commonsengine.governance.domain.Model.ProposalStatus;
import com.commonsengine.governance.domain.Model.ProposalType;
import com.commonsengine.governance.domain.Model.StakeholderType;
import com.commonsengine.governance.domain.Model.Vote;
import com.commonsengine.governance.domain.Model.VoteChoice;
import com.commonsengine.governance.domain.Model.VoteResult;
import com.commonsengine.governance.infrastructure.persistence.GovernancePersistence.ProposalEntity;
import com.commonsengine.governance.infrastructure.persistence.GovernancePersistence.VoteEntity;
import com.commonsengine.governance.infrastructure.persistence.GovernancePersistence.ProposalMapper;
import com.commonsengine.governance.infrastructure.persistence.GovernancePersistence.VoteMapper;
import com.commonsengine.governance.infrastructure.persistence.GovernanceRepositories.ProposalRepository;
import com.commonsengine.governance.infrastructure.persistence.GovernanceRepositories.VoteRepository;
import com.commonsengine.platform.exception.BusinessRuleException;
import com.commonsengine.platform.exception.NotFoundException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Duration;
import java.time.Instant;
import java.util.List;

/**
 * 治理业务服务（#69 治理模块）
 *
 * 职责：
 * 1. 创建提案（讨论期 → 投票期 → 决议）
 * 2. 开启投票（讨论期满后）
 * 3. 投票（投票期内，一人一票）
 * 4. 计票（加权：WORKER 0.40 / CONSUMER 0.30 / COMMUNITY 0.30）
 * 5. 关闭投票并决议
 *
 * 治理原则（对齐架构文档「民主治理 1.4」）：
 * - 利益相关者加权投票，非简单多数
 * - 提案有强制讨论期，防止冲动决议
 */
@Service
public class GovernanceService {

    private static final Logger log = LoggerFactory.getLogger(GovernanceService.class);

    /** 默认讨论期（小时） */
    private static final long DEFAULT_DISCUSSION_HOURS = 72;
    /** 默认投票期（小时） */
    private static final long DEFAULT_VOTING_HOURS = 48;

    private final ProposalRepository proposalRepository;
    private final VoteRepository voteRepository;

    public GovernanceService(ProposalRepository proposalRepository,
                              VoteRepository voteRepository) {
        this.proposalRepository = proposalRepository;
        this.voteRepository = voteRepository;
    }

    @Transactional
    public Proposal createProposal(String title, String description, ProposalType type,
                                    String proposerId) {
        Instant now = Instant.now();
        Proposal proposal = new Proposal(
                ProposalId.random(),
                title,
                description,
                type,
                proposerId,
                ProposalStatus.DISCUSSION,
                now,
                now.plus(Duration.ofHours(DEFAULT_DISCUSSION_HOURS)),
                now.plus(Duration.ofHours(DEFAULT_DISCUSSION_HOURS + DEFAULT_VOTING_HOURS)),
                DEFAULT_DISCUSSION_HOURS,
                DEFAULT_VOTING_HOURS
        );
        ProposalEntity saved = proposalRepository.save(ProposalMapper.toEntity(proposal));
        log.info("提案已创建 id={} type={}", proposal.id(), type);
        return ProposalMapper.toDomain(saved);
    }

    /**
     * 投票——必须在投票期内，且一人一票。
     */
    @Transactional
    public Vote castVote(ProposalId proposalId, String voterId,
                          StakeholderType stakeholderType, VoteChoice choice, String reason) {
        Proposal proposal = loadProposalOrThrow(proposalId);
        Instant now = Instant.now();

        if (proposal.status() != ProposalStatus.VOTING) {
            if (proposal.status() == ProposalStatus.DISCUSSION) {
                throw new BusinessRuleException(
                        "PROPOSAL_STILL_IN_DISCUSSION",
                        "提案仍在讨论期，不能投票。投票开放时间: " + proposal.votingOpensAt()
                );
            }
            throw new BusinessRuleException(
                    "PROPOSAL_NOT_VOTABLE",
                    "提案状态为 " + proposal.status() + "，不接受投票"
            );
        }
        if (!proposal.isVotingOpen(now)) {
            throw new BusinessRuleException(
                    "VOTING_CLOSED",
                    "投票期已结束"
            );
        }

        // 一人一票校验
        if (!voteRepository.findByProposalIdAndVoterId(proposalId.value(), voterId).isEmpty()) {
            throw new BusinessRuleException(
                    "ALREADY_VOTED",
                    "投票者 " + voterId + " 已对此提案投过票"
            );
        }

        Vote vote = Vote.create(proposalId, voterId, stakeholderType, choice, reason);
        VoteEntity saved = voteRepository.save(VoteMapper.toEntity(vote));
        log.info("投票已记录 proposal={} voter={} choice={}", proposalId, voterId, choice);
        return VoteMapper.toDomain(saved);
    }

    /**
     * 开启投票——讨论期满后手动触发，或自动在计票时检查。
     */
    @Transactional
    public Proposal openVoting(ProposalId proposalId) {
        Proposal proposal = loadProposalOrThrow(proposalId);
        if (proposal.status() != ProposalStatus.DISCUSSION) {
            throw new BusinessRuleException(
                    "PROPOSAL_NOT_IN_DISCUSSION",
                    "提案状态为 " + proposal.status() + "，不在讨论期"
            );
        }
        if (!proposal.isDiscussionOver(Instant.now())) {
            throw new BusinessRuleException(
                    "DISCUSSION_NOT_OVER",
                    "讨论期未满，投票开放时间: " + proposal.votingOpensAt()
            );
        }

        Proposal updated = new Proposal(
                proposal.id(), proposal.title(), proposal.description(), proposal.type(),
                proposal.proposerId(), ProposalStatus.VOTING, proposal.createdAt(),
                proposal.votingOpensAt(), proposal.votingClosesAt(),
                proposal.discussionDurationHours(), proposal.votingDurationHours()
        );
        ProposalEntity saved = proposalRepository.save(ProposalMapper.toEntity(updated));
        log.info("提案进入投票期 id={}", proposalId);
        return ProposalMapper.toDomain(saved);
    }

    /**
     * 计票并决议——投票期结束后触发。
     *
     * 加权计票：APPROVE +权重，REJECT −权重，ABSTAIN 不计。
     * 最终得分 > 0 视为通过。
     */
    @Transactional
    public VoteResult tallyAndResolve(ProposalId proposalId) {
        Proposal proposal = loadProposalOrThrow(proposalId);
        if (proposal.status() != ProposalStatus.VOTING) {
            throw new BusinessRuleException(
                    "PROPOSAL_NOT_IN_VOTING",
                    "提案状态为 " + proposal.status() + "，不在投票期"
            );
        }

        List<VoteEntity> voteEntities = voteRepository.findByProposalId(proposalId.value());
        int approve = 0, reject = 0, abstain = 0;
        double weightedScore = 0;

        for (VoteEntity ve : voteEntities) {
            double weight = ve.getStakeholderType().getWeight();
            switch (ve.getChoice()) {
                case APPROVE -> {
                    approve++;
                    weightedScore += weight;
                }
                case REJECT -> {
                    reject++;
                    weightedScore -= weight;
                }
                case ABSTAIN -> abstain++;
            }
        }

        boolean passed = weightedScore > 0;
        ProposalStatus finalStatus = passed ? ProposalStatus.PASSED : ProposalStatus.REJECTED;

        Proposal resolved = new Proposal(
                proposal.id(), proposal.title(), proposal.description(), proposal.type(),
                proposal.proposerId(), finalStatus, proposal.createdAt(),
                proposal.votingOpensAt(), proposal.votingClosesAt(),
                proposal.discussionDurationHours(), proposal.votingDurationHours()
        );
        proposalRepository.save(ProposalMapper.toEntity(resolved));

        VoteResult result = new VoteResult(
                proposalId, voteEntities.size(), approve, reject, abstain, weightedScore, passed
        );
        log.info("提案计票完成 id={} total={} approve={} reject={} abstain={} score={} passed={}",
                proposalId, voteEntities.size(), approve, reject, abstain, weightedScore, passed);
        return result;
    }

    @Transactional(readOnly = true)
    public Proposal findById(ProposalId id) {
        return loadProposalOrThrow(id);
    }

    @Transactional(readOnly = true)
    public List<Proposal> findByStatus(ProposalStatus status) {
        return proposalRepository.findByStatusOrderByCreatedAtDesc(status).stream()
                .map(ProposalMapper::toDomain)
                .toList();
    }

    @Transactional(readOnly = true)
    public List<Vote> findVotes(ProposalId proposalId) {
        return voteRepository.findByProposalId(proposalId.value()).stream()
                .map(VoteMapper::toDomain)
                .toList();
    }

    private Proposal loadProposalOrThrow(ProposalId id) {
        return proposalRepository.findById(id.value())
                .map(ProposalMapper::toDomain)
                .orElseThrow(() -> new NotFoundException("Proposal", id.value()));
    }
}

File: backend/governance/src/main/kotlin/com/commonsengine/governance/service/GovernanceService.kt
package com.commonsengine.governance.service

import com.commonsengine.governance.domain.Proposal
import com.commonsengine.governance.domain.ProposalId
import com.commonsengine.governance.domain.ProposalStatus
import com.commonsengine.governance.domain.ProposalType
import com.commonsengine.governance.domain.StakeholderType
import com.commonsengine.governance.domain.Vote
import com.commonsengine.governance.domain.VoteChoice
import com.commonsengine.governance.domain.VoteResult
import com.commonsengine.governance.domain.weight
import com.commonsengine.governance.infrastructure.persistence.ProposalRepository
import com.commonsengine.governance.infrastructure.persistence.VoteRepository
import com.commonsengine.governance.infrastructure.persistence.toDomain
import com.commonsengine.governance.infrastructure.persistence.toEntity
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.Instant
import java.time.temporal.ChronoUnit

/**
 * 治理服务——合作社民主治理的技术载体
 *
 * 核心原则：
 * 1. 多利益相关方加权投票（劳动者 40% / 消费者 30% / 社区 30%）
 * 2. 提案至少讨论 30 天后才能投票
 * 3. 章程修改需 2/3 多数 + 45 天讨论
 * 4. 涉及不可篡改底线的提案无效
 *
 * 注意：治理参数的最终值需由全体大会决定，本服务只做技术载体。
 *
 * 持久化：使用 JPA + PostgreSQL，重启不丢数据。
 */
@Service
open class GovernanceService(
    private val proposalRepository: ProposalRepository,
    private val voteRepository: VoteRepository,
) {

    /** 提交提案 */
    @Transactional
    open fun createProposal(
        title: String,
        description: String,
        proposedBy: String,
        type: ProposalType = ProposalType.OTHER,
    ): Proposal {
        require(title.isNotBlank()) { "提案标题不能为空" }

        val discussionDays = if (type == ProposalType.CHARTER_AMENDMENT) 45 else 30

        val proposal = Proposal(
            id = ProposalId.random(),
            title = title,
            description = description,
            proposedBy = proposedBy,
            type = type,
            discussionDeadline = Instant.now().plus(discussionDays.toLong(), ChronoUnit.DAYS),
        )
        proposalRepository.save(proposal.toEntity())
        return proposal
    }

    /** 开启投票（讨论期结束后） */
    @Transactional
    open fun startVote(proposalId: ProposalId): Proposal? {
        val entity = proposalRepository.findByProposalId(proposalId.value)
            ?: return null
        val proposal = entity.toDomain()

        require(proposal.status == ProposalStatus.DISCUSSION) { "提案必须处于讨论状态" }
        require(Instant.now().isAfter(proposal.discussionDeadline)) {
            "讨论期未满，不能开始投票（截止: ${proposal.discussionDeadline}）"
        }

        entity.status = ProposalStatus.VOTING
        proposalRepository.save(entity)
        return entity.toDomain()
    }

    /** 投票 */
    @Transactional
    open fun castVote(
        proposalId: ProposalId,
        voterId: String,
        stakeholderType: StakeholderType,
        choice: VoteChoice,
    ): Vote {
        val entity = proposalRepository.findByProposalId(proposalId.value)
            ?: throw IllegalArgumentException("提案不存在: $proposalId")
        val proposal = entity.toDomain()

        require(proposal.status == ProposalStatus.VOTING) { "提案不在投票阶段" }

        // 检查是否已投票（一人一票）——数据库唯一约束保障
        require(!voteRepository.existsByProposalIdAndVoterId(proposalId.value, voterId)) {
            "$voterId 已对此提案投过票"
        }

        val vote = Vote(
            proposalId = proposalId,
            voterId = voterId,
            stakeholderType = stakeholderType,
            choice = choice,
        )
        voteRepository.save(vote.toEntity())
        return vote
    }

    /** 统计投票结果 */
    @Transactional
    open fun tallyVotes(proposalId: ProposalId): VoteResult {
        val proposalVotes = voteRepository.findByProposalId(proposalId.value).map { it.toDomain() }
        val entity = proposalRepository.findByProposalId(proposalId.value)
        val proposal = entity?.toDomain()

        val yesWeight = proposalVotes.filter { it.choice == VoteChoice.YES }
            .sumOf { it.stakeholderType.weight }
        val noWeight = proposalVotes.filter { it.choice == VoteChoice.NO }
            .sumOf { it.stakeholderType.weight }
        val abstainCount = proposalVotes.count { it.choice == VoteChoice.ABSTAIN }

        // 通过门槛：普通提案简单多数，章程修改需 2/3
        val threshold = if (proposal?.type == ProposalType.CHARTER_AMENDMENT) 2.0 / 3.0 else 0.5
        val totalValidWeight = yesWeight + noWeight
        val passed = if (totalValidWeight > 0) yesWeight / totalValidWeight > threshold else false

        // 更新提案状态
        if (entity != null && entity.status == ProposalStatus.VOTING) {
            entity.status = if (passed) ProposalStatus.APPROVED else ProposalStatus.REJECTED
            proposalRepository.save(entity)
        }

        return VoteResult(
            proposalId = proposalId,
            yesWeighted = yesWeight,
            noWeighted = noWeight,
            abstainCount = abstainCount,
            totalVotes = proposalVotes.size,
            passed = passed,
            breakdown = buildString {
                append("投票结果：赞成 ${"%.1f%%".format(yesWeight * 100)}（加权）/ ")
                append("反对 ${"%.1f%%".format(noWeight * 100)}（加权）/ ")
                append("弃权 $abstainCount 票。")
                append("通过门槛：${"%.0f%%".format(threshold * 100)}。")
                append(if (passed) "提案通过。" else "提案未通过。")
            },
        )
    }

    /** 查询所有提案 */
    @Transactional(readOnly = true)
    open fun findAllProposals(): List<Proposal> =
        proposalRepository.findAll().map { it.toDomain() }

    /** 查询提案详情 */
    @Transactional(readOnly = true)
    open fun findProposal(id: ProposalId): Proposal? =
        proposalRepository.findByProposalId(id.value)?.toDomain()
}
